base/iobuffer.jl

```diff
@@ -317,118 +484,144 @@ function seek(io::GenericIOBuffer, n::Int)
     return io
 end
 
+# TODO: Should check for seekable and error if not
 function seekend(io::GenericIOBuffer)
     io.ptr = io.size+1
     return io
 end
 
-# choose a resize strategy based on whether `resize!` is defined:
-# for a Vector, we use `resize!`, but for most other types,
-# this calls `similar`+copy
-function _resize!(io::GenericIOBuffer, sz::Int)
-    a = io.data
-    offset = io.offset
-    if applicable(resize!, a, sz)
-        if offset != 0
-            size = io.size
-            size > offset && copyto!(a, 1, a, offset + 1, min(sz, size - offset))
-            io.ptr -= offset
-            io.size -= offset
-            io.offset = 0
-        end
-        resize!(a, sz)
+# Resize data to exactly size `sz`. Either resize the underlying data,
+# or allocate a new one and copy.
+# This should only be called after the offset is zero - any operation which calls
+# _resize! should reset offset before so.
+function _resize!(io::GenericIOBuffer, new_size::Int)
+    old_data = io.data
+    if applicable(resize!, old_data, new_size)
+        resize!(old_data, new_size)
     else
         size = io.size
-        if size >= sz && sz != 0
-            b = a
+        # Make a new data buffer, only if there is not room in existing buffer
+        if size >= new_size && !iszero(new_size)
+            new_data = old_data
         else
-            b = _similar_data(io, sz == 0 ? 0 : max(overallocation(size - io.offset), sz))
+            new_data = _similar_data(io, new_size)
+            io.data = new_data
         end
-        size > offset && copyto!(b, 1, a, offset + 1, min(sz, size - offset))
-        io.data = b
-        io.ptr -= offset
-        io.size -= offset
-        io.offset = 0
+        size > 0 && copyto!(new_data, 1, old_data, 1, min(new_size, size))
     end
     return io
 end
 
+# TODO: These errors cannot be converted to LazyString, but it's wasteful to interpolate them here.
```

Comment:
It is generally best not to have comments saying something cannot be done, since they have a tendency to become stale, and then prevent someone else from considering doing the very thing the comment wanted to do
```suggestion
```